### DISCUSSION TITLE
Which (small) issues should Flock fix first?

### DISCUSSION THREAD

**Discussion by matthew-carroll** [join-the-flock/nest#24] [Q&A]
Is there a (small) issue that's driving you nuts in Flutter? Do you think Flock should make that issue one of it's first bug fixes? If so, post a link to the Flutter issue ticket below, **and provide a brief description about why it should be a priority.**

**Comment by orestesgaolin** [comment]
For me it would be the scrolling fidelity on iOS. As a primarily mobile app developer I often face the challenge of not being able to reproduce familiar scrolling behavior with Flutter. Not even mentioning issue with keyboard dismissal, there are just physics differences:
https://github.com/flutter/flutter/issues/149221
https://github.com/flutter/flutter/issues/142441

and the keyboard behavior:
https://github.com/flutter/flutter/issues/57609

**Reply by matthew-carroll** [reply]
Quick note on this one, I think that probably 99% of the challenge is finding/assembling a momentum equation that feels nearly identical to iOS. The Flutter team put a lot of effort into that, but apparently came up short. If you or others have ideas about how to do better, then it could be a nice win for Flock.

**Reply by matthew-carroll** [reply]
@jezell do you think the keyboard synchronization with scrolling might fundamentally be a threading issue? And maybe merging threads on iOS will at least make this possible? https://github.com/flutter/flutter/issues/57609

**Reply by jezell** [reply]
@matthew-carroll The comments on these threads seem to suggest the scrolling differences are physics related. My gut would be that the keyboard behavior difference probably isn't totally threading related since the keyboard is a different process on iOS, but definitely minor latency could be introduced by the current setup.

**Comment by EArminjon** [comment]
Hello :)

By small issue you mean an issue where a PR is open or issue where fix must not be so hard to achieve ?

**Reply by matthew-carroll** [reply]
The later - an issue where the fix isn't too hard (and also doesn't change a large number of lines of code)

**Reply by EArminjon** [reply]
Understood.

I think about this one for iOS : https://github.com/flutter/flutter/issues/33875
Swipe to back can freeze an app in some scenario.

**Reply by matthew-carroll** [reply]
Can you describe why this is a priority?

**Reply by EArminjon** [reply]
The priority depend if an app contain the issue scenario (ex: show flushbars). For these apps, the freeze can't be fixed and require a restart when it happens. Not the biggest priority for 'normal' apps but painful one.

**Comment by NabilaWorks** [comment]
The one where Flutter somehow adds an extra frame of latency on iOS would be wonderful : https://github.com/flutter/flutter/issues/110431

Easiest one would probably be adding the '@' keyboard type on iOS https://github.com/flutter/flutter/issues/64801

**Reply by matthew-carroll** [reply]
Please include an explanation for why this should be a community priority.

**Reply by NabilaWorks** [reply]
Amongst all the suggested issues as of now, https://github.com/flutter/flutter/issues/110431 has the most thumbs up / reactions.

A proper fix in Flock has the potential to provide tangible value to all of those participants.

It's likely an iOS only issue ( which narrows down the scope ) and @knopp was already able to do a quick and dirty hack himself to trigger beginning a frame immediately after an event (instead of scheduling vsync).

Since Flutter seems to skip an entire frame when handling touch events, this affects every single iOS Flutter app.

Every Flutter developer who maintains an iOS app would be able to try Flock and immediately improve the responsiveness in their app if Flock contained this fix.

**Reply by matthew-carroll** [reply]
Thanks for elaborating on scope of who's effected. How about the level of impact per user? What's the average negative impact of being one frame behind on gestures? Even reading the issue, it's defined as a technical problem. I didn't see any user impact description in the issue ticket.

@knopp do you have any thoughts on the user impact and/or the fixing this in Flock?

**Comment by alexmercerind** [comment]
Smooth scrolling with mouse wheel on desktop: https://github.com/flutter/flutter/issues/32120

Appears to be a framework (non engine) only issue.

**Reply by matthew-carroll** [reply]
Please include an explanation for why this should be a community priority.

**Comment by temaand1** [comment]
This one is super annoying
https://github.com/flutter/flutter/issues/120353

**Reply by matthew-carroll** [reply]
Please include an explanation for why this should be a community priority.

**Comment by liweilong** [comment]
https://github.com/flutter/flutter/issues/63043
text rendering with very bad anti-aliasing have been a annoying problem for a very long time.

**Reply by matthew-carroll** [reply]
Please include an explanation for why this should be a community priority.

**Comment by Uvais-Mohammad** [comment]
Popscope is not working properly with GoRouter. [Issue Link](https://github.com/flutter/flutter/issues/138737). 
GoRouter is a popular package for declarative routing in Flutter, it supports deeplinking.

**Comment by matthew-carroll** [comment]
Apple Magic Mouse touch events don't respect native behaviors. The following ticket identifies a scroll followed by a tap as a failure. There may be other common Mac interactions that are also unsupported.

The fix involves at least some nuanced rework of built-in gesture recognizers, and perhaps passing some new infro from the platform.

https://github.com/flutter/flutter/issues/120788

CC @knopp

**Comment by dkbast** [comment]
https://github.com/flutter/flutter/issues/138624 
PopScope on ios is broken with showModalBottomSheet - I want to execute a callback when trying to dismiss a bottom sheet which is set as "non dismissible", the issue above was closed for it seems this behaviour is intended in general navigation, but it should not be for bottom sheets. Could be a quick win?

**Comment by matthew-carroll** [comment]
Here are some issues that we've discovered at Super Editor, which impact general Flutter app developers.

### Image rendering
`RenderImage` constantly runs layout when playing: https://github.com/flutter/flutter/issues/123441

### Initial route
The timing of the initial route (AKA deep links) is delayed. It should be sent immediately upon app launch: https://github.com/flutter/flutter/issues/146174

### Bugs in text input for various languages
**Portuguese**
 * Compound character creates extra character: https://github.com/flutter/flutter/issues/156184
 * Pressing `´` deletes character: https://github.com/flutter/flutter/issues/156183
 * Typing compound character generates unexpected deltas on Safari: https://github.com/flutter/flutter/issues/134092

**Japanese**
 * IME suggestion panel duplicates the text that's confirmed: https://github.com/flutter/flutter/issues/153065
 * Invalid character is inserted after selecting suggested text: https://github.com/flutter/flutter/issues/142493

**Korean**
 * Bad deletion range when IME replaces a character: https://github.com/flutter/flutter/issues/142894
 * Unexpected deltas on Web on Windows: https://github.com/flutter/flutter/issues/139603

**Any Language**
Backspacing a composing character requires pressing backspace twice: https://github.com/flutter/flutter/issues/104950

### Text Layout
 * Text height when going from empty to non-empty is inconsistent: https://github.com/flutter/flutter/issues/155620
 * Text caret height is wrong when queried for a trailing space: https://github.com/flutter/flutter/issues/155330
 * `RenderParagraph` should call `markNeedsLayout` when `textAlign` is changed: https://github.com/flutter/flutter/issues/140756
 * Non-breaking space character isn't honored when using `WidgetSpan`s: https://github.com/flutter/flutter/issues/148259
 * Superscript text doesn't begin with the right character: https://github.com/flutter/flutter/issues/150744
 * Superscript and subscript don't work on web: https://github.com/flutter/flutter/issues/150743

### OS Popovers and System Text Input
 * Provide a way to close Mac accent panel: https://github.com/flutter/flutter/issues/132272
 * Dismissing the Mac accent panel with ESC produces an invalid character: https://github.com/flutter/flutter/issues/153895
 * Position emoji picker on Web: https://github.com/flutter/flutter/issues/134265
 * Unable to detect if native composing panel is open on Web: https://github.com/flutter/flutter/issues/134268
 * Can't disable the context menu on Mobile Web: https://github.com/flutter/flutter/issues/134358
 * Add support for Mac OS text replacements: https://github.com/flutter/flutter/issues/145122
 * Newline is reported at end of line when it shouldn't be: https://github.com/flutter/flutter/issues/148604

**Reply by jezell** [reply]
Great list

**Reply by Liloupar** [reply]
and this
https://github.com/flutter/flutter/issues/115257

**Reply by delfme** [reply]
@matthew-carroll @jezell Two of Text's most severe bugs are still missing. Could you please add these to the list as well?

https://github.com/flutter/flutter/issues/28894#issuecomment-1537587062

https://github.com/flutter/flutter/issues/120857

**Comment by simonbengtsson** [comment]
Supporting leading space for TextSpan would be awesome https://github.com/flutter/flutter/issues/128218. That is pretty much the only thing blocking using the native TextField to create a basic rich text field. Pretty much everything such as text formatting (bold, code blocks etc), keyboard shortcuts, links, markdown text shortcuts etc is possible but bullet and number lists is not really feasible due to above not being supported.

**Reply by matthew-carroll** [reply]
It looks like the linked ticket is trying to turn a text field into a text editor. Is there a reason that you don't use an actual text editor? For example, Super Editor?

Reading through that ticket it seems to me that the root problem is that a text field is being used for something well beyond it's intended purpose, so you're running into predictable difficulties.

**Reply by simonbengtsson** [reply]
That might be true. Was a few months ago I last tried super editor, but at that time I was unable to get dark mode, undo/redo, keyboard behavior on mobile etc working as well as what was supported out of the box with the native text field.

**Reply by matthew-carroll** [reply]
If you haven't already, it would be helpful if you could file tickets with Super Editor for the things you believe are lacking.

WRT "the native text field" - my point above was that the concept of creating indented blocks of text, or any other blocks, goes beyond what a text field exists to do. This box that Im typing into right now might look kind of like a text field, but in reality, it's a document editor. If you want multiple paragraphs with any kind of block-level formatting, then I think what you want is a document editor, not a text field.

**Reply by matthew-carroll** [reply]
BTW, the reason this is relevant to your original post is because Flutter doesn't ship a document editor, nor should it. So if what you're looking for is document editing, then that doesn't really belong in Flutter.

**Comment by softshape** [comment]
CameraX in the original Flutter is so buggy that our users suffer so much from sudden app crashes. If you're able to fix the camera plugin, that would move us to replace Flutter with Flock.

**Comment by fonkamloic** [comment]
Sorry, this is not considered a small issue. And so disregard this comment if it can't be considered. 

However, in my opinion, this is the most relevant and impactful issue to solve. [Enable "hot reload" (not just "hot restart") for FlutterWeb](https://github.com/flutter/flutter/issues/53041)

Hot reload in Mobile Flutter is one of the major reasons Flutter is so appealing and inviting even to native app developers. 

I know the company I work for will use Flock in production for all its 10+ enterprise clients if FlutterWeb hot reload is implemented.

**Reply by matthew-carroll** [reply]
@timu-jesse-ezell

**Comment by J3rr1ck** [comment]
Hey all so not a ton of quick fixes for the linux platform at the moment but a quicker patch I'm seeing that could be potentially useful to users I know. 
[https://github.com/flutter/flutter/issues/159460)](https://github.com/flutter/flutter/issues/159460)
Bit more complex that I'd typically pick for a early fix but I sense attention to bugs like these may be a good way to set ourselves apart from using mainline flutter on linux.
Further justifications from the flutter community can be found here: 
[https://github.com/flutter/flutter/issues/153560](https://github.com/flutter/flutter/issues/153560)

**Reply by matthew-carroll** [reply]
Do you believe this is something that you'll be able to technically lead? Is it reasonable to think we can make this switch, do it correctly, and maintain it?

**Reply by J3rr1ck** [reply]
Sorry about the slow response, I can definitely technically lead this as it's right in my wheel house, let me double check I'm not getting myself into fresh maintainer hell but I think it should be all a go on my end.

**Reply by matthew-carroll** [reply]
Great. Do you think this is something that should wait for Flutter to merge the engine into the framework? Or does it make sense to get started on this before that?

**Reply by J3rr1ck** [reply]
that's a great point I hadn't fully taken into consideration, may be the wisest to keep this on the back burner till the merge to prevent craziness down the line. Do we know what the timeline for that process is currently looking like for merge?

**Reply by matthew-carroll** [reply]
Should be done by the end of this month

**Comment by wer-mathurin** [comment]
Simple to resolve with a small fix.
https://github.com/flutter/flutter/issues/102861

As described here: When you fetch more data, you will see the first row of the next page displaying a CircularProgressIndicator  in each Cells, if the Column definition is not a numeric. This is BTW can look very strange if you have some column with numeric and other with not.

This is not respecting the material spécifications.

This looks like very ugly and not professional for enterprise app!

**Comment by rensPols** [comment]
This one is important I think, because the current state makes a Flutter app look crappy when the client uses a scroll wheel.
https://github.com/join-the-flock/nest/issues/25

I don't know for certain if this is easy to fix, but the packages that have been made to solve this are relatively simple. The problem with said packages is that they use the NeverScrollableScrollPhysics(), which makes the scrollbar unusable.

**Comment by azimuthdeveloper** [comment]
It would be easy to revert the nested heroes change. It broke several peoples apps when it landed. I'd suggest it as a community priority because you've got people modding the flutter framework files to manually remove the assert, so that kind of monkey patching is going to cause flakiness for sure.

Discussion here: https://github.com/flutter/flutter/issues/38684#issuecomment-2105880740

Revert PR https://github.com/flutter/flutter/pull/28470

**Reply by matthew-carroll** [reply]
The Flutter team marked is a P3 😂

But I'd like to understand more about what's going on here. The linked "Revert PR" is from 2019. The linked discussion about the problem is much more recent.

What's the history of this issue?

**Comment by fisforfaheem** [comment]


**Reply by matthew-carroll** [reply]
Please don't spam this. We're doing what we can.

**Comment by polonski** [comment]
Mildly annoying, and hopefully quick to fix: Way to disable the [yellow-and-black debug overflow indicator](https://github.com/join-the-flock/flock/blob/master/packages/flutter/lib/src/rendering/debug_overflow_indicator.dart) during runtime. 
For example, when inputting some data into a form on a small resolution display, if it wasn't for this indicator, things would look ok, I don't care that the bottom keyboard overflows another component by 2 pixels.. go away please.

You could just change the [black and yellow colours](https://github.com/flutter/flutter/blob/main/packages/flutter/lib/src/rendering/debug_overflow_indicator.dart#L90-L91) to alpha 0 , but there's probably a better way of doing this.

I know that the solution in Flutter is to use another Container to wrap things and prevent the overflow indicator, but I don't mind if it does.
 
Also what if want text to overflow on purpose? (for art)

**Reply by matthew-carroll** [reply]
I'm not sure what specific cases you're referring to, but overflow is an error. There are widgets available to allow for overflow, such as `OverflowBox`. You can also write custom render objects to handle layout that's otherwise too tricky to do with existing widgets. And if you want to paint outside your layout bounds, you can implement a CustomPainter, which can paint the entire case if you want it to. 

If an overflow is happening within Flutter widgets when it shouldn't, then that sounds like a bug in Flutter that you should file. 

But disabling a notice for a legitimate error is unlikely to be something that we'd merge.

**Comment by azimuthdeveloper** [comment]
For some reason, there have been huge changes in how dart formats code. There's a huge amount written about it here: https://github.com/dart-lang/dart_style/issues/1253.

This change wasn't properly telegraphed and now the issue is lighting up with people who are discovering this formatting change and are asking for it to be reverted. So, if Flock shipped the language improvements with the latest Dart, but kept the old formatter, or made it configurable, that would be ideal. As it is, the new formatter is in use for new Flutter projects, and is not configurable by design. So you get what you get and you don't get upset.

**Comment by socvid** [comment]
We didn't notice this discussion, and we made an issue on nest instead: https://github.com/join-the-flock/nest/issues/50
Relevant Flutter issues are provided in this issue.

**Our request:**
We have successfully built the Flutter engine locally with Graphite support by including the Graphite source files and turning on the appropriate flags. We would happily make a PR to add this change to Flock if it seemed like it would get merged.

Is the Flock team open to this?

**Reply by matthew-carroll** [reply]
I'm open to it, with some important caveats that we can discuss later. @jezell how about you?

**Reply by jezell** [reply]
I think this is a great usecase. One of the things that's a bit annoying about Flutter is you only get the blessed build, rather than being able to turn on and off features depending on your needs. Some people really care about engine size, but I've always wanted a setup where people could pick and choose optional features during the engine / framework build process. Part of the reason the nest patch process is cool is because it can enable this type of scenario where you have an optional patch that turns on a specific feature if you need it.

**Reply by socvid** [reply]
Exactly, @jezell! This is really encouraging.

**Reply by mimbrown** [reply]
@matthew-carroll @jezell so just to clarify, you're open to a PR that would add a switch to enable Graphite support within Harfbuzz at build time, correct? How would you expect this to get into the hands of the end-developer? Is it basically, anyone can fork Flock, flip the feature switches they want, copy the github actions script, and get their own builds going?

**Reply by matthew-carroll** [reply]
Any such flag or integration needs to wait until we get the infrastructure deployed. Jesse is working on engine builds. I'm working on patch management. After that we can consider a contribution that integrates Graphite, and see if that makes sense.

**Reply by mimbrown** [reply]
Understood, thanks for your quick replies.

**Reply by jezell** [reply]
@mimbrown I'm on the tail end of it. Have the builds working, but need to make sure the CLI works for installing still. Hopefully everything is good to go this week. I'll ping you here when it's ready.

**Reply by socvid** [reply]
@jezell, how's the progress going?

**Reply by socvid** [reply]
@jezell @matthew-carroll, we're looking forward to seeing the CLI for engine builds and patch management up and running. Do you have an ETA?

**Reply by timu-jesse-ezell** [reply]
Was pushing up some PRs last week and will continue again this week. Really want to have this in place and ironed out before Fluttercon NYC.

**Reply by matthew-carroll** [reply]
I've been busy with some Fluttercon stuff and client work. I'll review @timu-jesse-ezell's work as soon as I can.

**Reply by socvid** [reply]
@jezell @matthew-carroll, looks like you got some significant PRs in before Fluttercon! That's great to see. Are there still some things to iron out or is patching Flutter with Flock ready for the public?

**Reply by matthew-carroll** [reply]
@socvid yes, we've been hacking on the patch system over time as the Flutter team has made changes to their own project structure. We think the Flutter changes are probably done, and we think we've identified a direction that's likely to work for us. Our patch work for `master` seems to be functional. @jezell and I need to go back and iron out some things with `beta` and `stable`. We also have a few final details to figure out with regard to building and hosting engine binaries.

**Reply by socvid** [reply]
@matthew-carroll thanks for the update! We believe in you guys! Keep us posted.

**Reply by socvid** [reply]
@matthew-carroll, how's the progress going? Are you building and hosting engine binaries yet?

**Reply by timu-jesse-ezell** [reply]
we don't build and host engine binaries currently because we don't have any engine patches. however, there is an action checked in which runs on the engine-builds branch and can be used build the engine binaries and upload them to a GCP bucket.

**Reply by socvid** [reply]
Ok then we'll start working on our fork, let's see if we can get this building!

**Reply by matthew-carroll** [reply]
FYI, I'm working on a desktop app to make it easier to manage the patches in any given fork. Whenever I get that shipped, you'll probably want to use it. It's proving to be a pain to try to manage patch files directly.

**Reply by socvid** [reply]
@matthew-carroll, thanks for the heads up and your work on this! Looking forward to its release!

**Reply by matthew-carroll** [reply]
@socvid and @timu-jesse-ezell would you two like to take this PR for a spin? This PR creates the Aviary app, which we'll use to manage patches. At the moment it's nothing more than a patch file visualizer for Flock, but it would be good to get some eye balls on it early.

Would be great to make sure both of you can build it, run it, and you're able to open Flock, which involves locally cloning the Flock patch repository behind the scenes on your machine.

I only generated a project for Mac OS because the toolbar design is based on the Mac OS traffic lights and their spacing. If either of you need Windows, then maybe you could contribute that deployment target and we could coordinate on how you can adapt the UI for Windows?

Here's a video of the current functionality:

https://github.com/user-attachments/assets/7d28bbe3-f561-4d40-8c51-a6b91fd5a2b1

Here's the PR:
https://github.com/join-the-flock/nest/pull/69

**Reply by socvid** [reply]
This is great! I am traveling over the next week, but @mimbrown might be able to get to this sooner. He has already started working with Flock.

**Reply by socvid** [reply]
@matthew-carroll, sorry that we haven't yet taken PR #69 for a spin. @mimbrown has been working on building flock with his own fork without our own patches, and he has run into some issues that he documented in #70. We figured we would try out Aviary once we are utilizing our own patches. We are very curious to hear feedback on #70 first. Thanks.

**Comment by mark-brin** [comment]
just wondering, why not fix it in the official Flutter repo and raise the PR? why even a fork of an already maintained repo?

**Reply by timu-jesse-ezell** [reply]
Because the working with the Flutter team is a waste of time.
